docs/integrations/streamlit.md

```diff
@@ -0,0 +1,194 @@
+            key="favorite",  # key to st.session_state
+        )
+
+if __name__ == "__main__":
+    app()
+```
+> ⚠ This example is illustratory and real applications quickly get more complex.
+
+
+## Hamilton + Streamlit
+Adding Hamilton to your Streamlit application can provide a better separation between the dataflow and the UI logic. They pair nicely together because Hamilton is also stateless. Once defined, each call to `Driver.execute()` is independent. Therefore, on each Streamlit rerun, you use `Driver.execute()` to complete computations. Using Hamilton this way allows you to write your dataflow into Python modules and outside of the Streamlit.
+
+### Example
+In this example, we will build a simple financial dashboard based on the Kaggle [Bank Marketing Dataset](https://www.kaggle.com/datasets/janiobachmann/bank-marketing-dataset).
+
+> The full code can be found on [GitHub](https://github.com/DAGWorks-Inc/hamilton/tree/main/examples/streamlit)
+
+First, Hamilton transformations are defined in the module `logic.py`. This includes downloading the data from the web, getting unique values for `job`, conducting groupby aggregates, and creating `plotly` figures.
+```python
+# logic.py
+import pandas as pd
+import plotly.express as px
+from plotly.graph_objs import Figure
+
+
+def base_df() -> pd.DataFrame:
+    path = "https://raw.githubusercontent.com/Lexie88rus/bank-marketing-analysis/master/bank.csv"
+    return pd.read_csv(path)
+
+
+def all_jobs(base_df: pd.DataFrame) -> list[str]:
+    return base_df["job"].unique().tolist()
+
+
+def balance_per_job(base_df: pd.DataFrame) -> pd.DataFrame:
+    return base_df.groupby("job")["balance"].describe().astype(int)
+
+
+def balance_per_job_boxplot(base_df: pd.DataFrame) -> Figure:
+    return px.box(base_df, x="job", y="balance")
+
+
+def job_df(base_df: pd.DataFrame, selected_job: str) -> pd.DataFrame:
+    return base_df.loc[base_df['job']==selected_job]
+
+
+def job_hist(job_df: pd.DataFrame) -> Figure:
+    return px.histogram(job_df["balance"])
+```
+
+
+Then, the Streamlit UI is defined in `app.py`. Notice a few things:
+- `app.py` doesn't have to depend on `pandas` and `plotly`.
+- `@cache_resource` allows to create the `Driver` only once.
+- `@cache_data` on `_execute()` will automatically cache any Hamilton result based on the combination of arguments (`final_vars`, `inputs`, and `overrides`)
+- `get_state_inputs()` and `get_state_overrides()` will collect values from user inputs.
+- `execute()` parses the inputs and overrides from the state and call `_execute()`.
+
+```python
+# app.py
+from typing import Optional
+
+from hamilton import driver
+import streamlit as st
+
+import logic
+
+# cache to avoid rebuilding the Driver
+@st.cache_resource
+def get_hamilton_driver() -> driver.Driver:
+    return (
+        driver.Builder()
+        .with_modules(logic)
+        .build()
+    )
+
+# cache results for the set of inputs
+@st.cache_data
```

Comment:
This is really cool, but I'm honestly surprised caching works with a dictionary. Quite nifty though! I guess we could also cache function-level data to...